rc/extra/go-tools.kak

```diff
@@ -0,0 +1,182 @@
+# - goimports for code formatting on save 
+# - gogetdoc for documentation display and source jump (needs jq) (github.com/zmb3/gogetdoc)
+# Needs the following tools in the path:
+# - jq for json deserializaton
+
+# Auto-completion
+# ‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾
+
+decl -hidden str go_complete_tmp_dir
+decl -hidden completions gocode_completions
+
+def go-complete -docstring "Complete the current selection with gocode" %{
+    %sh{
+        dir=$(mktemp -d -t kak-go.XXXXXXXX)
+        printf %s\\n "set buffer go_complete_tmp_dir ${dir}"
+        printf %s\\n "eval -no-hooks write ${dir}/buf"
+    }
+    %sh{
+        dir=${kak_opt_go_complete_tmp_dir}
+        (
+            gocode_data=$(gocode -f=godit --in=${dir}/buf autocomplete ${kak_cursor_byte_offset})
+            rm -r ${dir}
+            column_offset=$(echo "${gocode_data}" | head -n1 | cut -d, -f1)
+
+            header="${kak_cursor_line}.$((${kak_cursor_column} - $column_offset))@${kak_timestamp}"
+            compl=$(echo "${gocode_data}" | sed 1d | awk -F ",," '{print $2 "||" $1}' | paste -s -d: -)
+            printf %s\\n "eval -client '${kak_client}' %{
+                set buffer=${kak_bufname} gocode_completions '${header}:${compl}'
+            }" | kak -p ${kak_session}
+        ) > /dev/null 2>&1 < /dev/null &
+    }
+}
+
+def go-enable-autocomplete -docstring "Add gocode completion candidates to the completer" %{
+    set window completers "option=gocode_completions:%opt{completers}"
+    hook window -group go-autocomplete InsertIdle .* %{ try %{
+        exec -draft <a-h><a-k>[\w\.].\z<ret>
+        go-complete
+    } }
+    alias window complete go-complete
+}
+
+def go-disable-autocomplete -docstring "Disable gocode completion" %{
+    set window completers %sh{ printf %s\\n "'${kak_opt_completers}'" | sed 's/option=gocode_completions://g' }
+    rmhooks window go-autocomplete
+    unalias window complete go-complete
+}
+
+# Auto-format
+# ‾‾‾‾‾‾‾‾‾‾‾
+
+decl -hidden str go_format_tmp_dir
+
+def -params ..1 go-format \
+    -docstring "go-format [goimports]: custom formatter for go files" %{
+    %sh{
+        dir=$(mktemp -d -t kak-go.XXXXXXXX)
+        printf %s\\n "set buffer go_format_tmp_dir ${dir}"
+        printf %s\\n "eval -no-hooks write ${dir}/buf"
+    }
+    %sh{
+        dir=${kak_opt_go_format_tmp_dir}
+        if [ "$1" = "1" ]; then
+            fmt_cmd="goimports -srcdir '${kak_buffile}'"
+        else
+            fmt_cmd="gofmt -s"
+        fi
+        eval "${fmt_cmd} -e -w ${dir}/buf 2> ${dir}/stderr"
+        if [ $? ]; then
+            cp ${dir}/buf "${kak_buffile}"
+        else
+            # we should report error if linting isn't active
+            printf %s\\n "echo -debug '$(cat ${dir}/stderr)'"
+        fi
+        rm -r ${dir}
+    }
+    edit!
+}
+
+def go-enable-format-onsave -docstring "Enable formatting on save for go files" %{
+    hook buffer -group go-format-onsave BufWritePre .+\.go %{ go-format }
```

Comment:
Yeah, I'd rather let user write hooks because that make them discover Kakoune, hooks are a bit tricky but they allow for a good understanding of how Kakoune is designed.

Here for example, the user still needs to write a hook anyway, to call `go-enable-format-onsave` on the buffer (so maybe on BufSetOption). 

regarding `go-format-imports` you could just name use `-only-imports` or `-imports` in place of `1`.